File: ABC/ABC301-400/ABC364/E/DP/E.cpp
#include <bits/stdc++.h>
using namespace std;
const int inf = 1e9;

int main() {
	int N, X, Y;
	cin >> N >> X >> Y;
	vector<vector<int>> dp(N + 1, vector<int>(X + 1, inf));
	dp[0][0] = 0;
	for (int i = 0; i < N; i++) {
		int A, B;
		cin >> A >> B;
		for (int j = i; j >= 0; j--) {
			for (int k = X; k >= A; k--) {
				if (dp[j][k - A] < inf) {
					dp[j + 1][k] = min(dp[j + 1][k], dp[j][k - A] + B);
				}
			}
		}
	}
	for (int i = N; i >= 0; i--) {
		for (int j = 0; j <= X; j++) {
			if (dp[i][j] <= Y) {
				cout << min(i + 1, N) << '\n';
				return 0;
			}
		}
	}
}

File: ABC/ABC301-400/ABC364/E/DP/E.py
inf = 10 ** 9

N, X, Y = map(int, input().split())
dp = [[inf] * (X + 1) for _ in range(N + 1)]
dp[0][0] = 0
for i in range(N):
	A, B = map(int, input().split())
	for j in range(i, -1, -1):
		for k in range(X, A - 1, -1):
			if dp[j][k - A] < inf:
				dp[j + 1][k] = min(dp[j + 1][k], dp[j][k - A] + B)
for i in range(N, -1, -1):
	for j in range(X + 1):
		if dp[i][j] <= Y:
			print(min(i + 1, N))
			exit()